### PR TITLE
`Paywalls`: add `PaywallFooterViewController`

### PR DESCRIPTION
**Pull request by tonidero**
### Description
This adds a new `PaywallFooterViewController`. This VC will be used to render the footer view in the hybrids.

### CODE REVIEW

File: Tests/TestingApps/PaywallsTester/PaywallsTester/Views/AppContentView.swift [hunk 1]
```diff
@@ -158,8 +158,8 @@ struct AppContentView: View {
 
     #if !os(watchOS)
     private func presentPaywallViewController() {
-        let paywall = PaywallViewController(displayCloseButton: true)
-        paywall.modalPresentationStyle = .pageSheet
+        let paywall = PaywallFooterViewController()
+        paywall.modalPresentationStyle = .formSheet
```
Comment (tonidero) [inline Tests/TestingApps/PaywallsTester/PaywallsTester/Views/AppContentView.swift:162]: This is just me trying this. Will revert once I confirm it works as I expect

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import RevenueCat
+import SwiftUI
+import UIKit
+
+/// A view controller for displaying `PaywallData` for an `Offering`.
+///
+/// - Seealso: ``PaywallView`` for `SwiftUI`.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+@objc(RCPaywallFooterViewController)
+public final class PaywallFooterViewController: UIViewController {
```
Comment (NachoSoto) [inline RevenueCatUI/UIKit/PaywallFooterViewController.swift:26]: There's a lot of duplication here with `PaywallViewController`. I think we can implement this by subclassing `PaywallViewController` instead and overriding the `mode`.

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        self.hostingController.view.frame = self.view.bounds
+    }
+
+}
+
+/// Delegate for ``PaywallFooterViewController``.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+@objc(RCPaywallFooterViewControllerDelegate)
+public protocol PaywallFooterViewControllerDelegate: AnyObject {
```
Comment (NachoSoto) [inline RevenueCatUI/UIKit/PaywallFooterViewController.swift:93]: Then we don't need a new protocol and we can use the same one.

File: Tests/APITesters/RevenueCatUIAPITester/SwiftAPITester/PaywallViewControllerAPI.swift [hunk 4]
```diff
@@ -19,6 +19,16 @@ func paywallViewControllerAPI(_ delegate: Delegate, _ offering: Offering?) {
     let _: UIViewController = PaywallViewController(offering: offering, displayCloseButton: true)
 }
 
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
+func paywallFooterViewControllerAPI(_ delegate: Delegate, _ offering: Offering?) {
+    let controller = PaywallFooterViewController()
+    controller.delegate = delegate
+
+    let _: UIViewController = PaywallFooterViewController(offering: offering)
+    let _: UIViewController = PaywallFooterViewController(displayCloseButton: true)
```
Comment (tonidero) [inline Tests/APITesters/RevenueCatUIAPITester/SwiftAPITester/PaywallViewControllerAPI.swift:28]: One of the consequences of inheriting from `PaywallViewController` is that we get this constructor, which shouldn't exist in the footer version...

**Comment by tonidero** [comment]
Will continue with this after Hack week. The current state is that I was trying to find a way to constraint the footer to just take the minimum required space before jumping to use it in the hybrids, since I think that would be required for that to work

File: Tests/APITesters/RevenueCatUIAPITester/SwiftAPITester/PaywallViewControllerAPI.swift (hunk 4 above)
Comment (NachoSoto) [reply Tests/APITesters/RevenueCatUIAPITester/SwiftAPITester/PaywallViewControllerAPI.swift:28]: You can override it and mark it unavailable:
```
@available(*, unavailable)
override init(displayCloseButton: Bool){}
```

File: RevenueCatUI/UIKit/PaywallFooterViewController.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+//  Created by Antonio Rico Diez on 4/12/23.
+//
+
+#if canImport(UIKit) && !os(tvOS) && !os(watchOS)
+
+import RevenueCat
+import UIKit
+
+/// A view controller for displaying `PaywallData` for an `Offering` in a footer format.
+/// This is used by the RevenueCathybrid SDKs in order to get a View of the footer.
```
Comment (NachoSoto) [inline RevenueCatUI/UIKit/PaywallFooterViewController.swift:21]:
```suggestion
/// This is used by the RevenueCat hybrid SDKs in order to get a View of the footer.
```

File: RevenueCatUI/UIKit/PaywallViewController.swift [hunk 6]
```diff
@@ -51,9 +55,12 @@ public final class PaywallViewController: UIViewController {
     }
 
     private lazy var hostingController: UIHostingController<some View> = {
-        let paywallView = self.offering
-            .map { PaywallView(offering: $0, displayCloseButton: self.displayCloseButton) }
-            ?? PaywallView(displayCloseButton: self.displayCloseButton)
+        let paywallView = PaywallView(offering: self.offering,
```
Comment (NachoSoto) [inline RevenueCatUI/UIKit/PaywallViewController.swift:58]: We can inline this now.

File: RevenueCatUI/UIKit/PaywallViewController.swift [hunk 7]
```diff
@@ -22,11 +22,15 @@ import UIKit
 /// - Seealso: ``PaywallView`` for `SwiftUI`.
 @available(iOS 15.0, macOS 12.0, tvOS 15.0, *)
 @objc(RCPaywallViewController)
-public final class PaywallViewController: UIViewController {
+public class PaywallViewController: UIViewController {
 
     /// See ``PaywallViewControllerDelegate`` for receiving purchase events.
     public weak var delegate: PaywallViewControllerDelegate?
```
Comment (NachoSoto) [inline RevenueCatUI/UIKit/PaywallViewController.swift:28]: Let's make this `final`

**Review by NachoSoto — COMMENTED** [review]
@tonidero this should be mostly ready, right?

**Comment by tonidero** [comment]
> @tonidero this should be mostly ready, right?

My only doubt right now is whether we need to add constraints for this to use the minimum needed space, or whether that would be in the hybrid side... But I think we can add that later if needed.

**Review by NachoSoto — APPROVED** [review]

**Pull request merged**